File: stacks/opinionated/laravel-app/app/Http/Controllers/ProductsController.php
<?php

namespace App\Http\Controllers;

use App\Models\Product;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Validator;

class ProductsController extends Controller
{
    public function store(Request $request): JsonResponse
    {
        $validator = Validator::make($request->all(), [
            'name' => ['required', 'string'],
            'price' => ['required', 'numeric', 'gt:0'],
        ]);

        if ($validator->fails()) {
            return response()->json(['detail' => $validator->errors()->first()], 422);
        }

        $validated = $validator->validated();
        $product = Product::create($validated);

        return response()->json($product, 201);
    }

    public function index(Request $request): JsonResponse
    {
        $query = Product::query()->orderBy('id');

        if ($request->has('min_price')) {
            if (! is_numeric($request->input('min_price')) || (float) $request->input('min_price') <= 0) {
                return response()->json(['detail' => "Query parameter 'min_price' is invalid"], 422);
            }
            $query->where('price', '>=', (float) $request->input('min_price'));
        }

        if ($request->has('max_price')) {
            if (! is_numeric($request->input('max_price')) || (float) $request->input('max_price') <= 0) {
                return response()->json(['detail' => "Query parameter 'max_price' is invalid"], 422);
            }
            $query->where('price', '<=', (float) $request->input('max_price'));
        }

        return response()->json($query->get());
    }

    public function show(int $id): JsonResponse
    {
        $product = Product::query()->find($id);

        if (! $product) {
            return response()->json(['detail' => 'Product not found'], 404);
        }

        return response()->json($product);
    }
}

File: stacks/opinionated/rails-app/app/controllers/products_controller.rb
class ProductsController < ApplicationController
  def create
    product = Product.create!(params.permit(:name, :price))
    render json: serialize_product(product), status: :created
  end

  def index
    products = Product.order(:id)
    min_price = numeric_query_param(:min_price)
    return if performed?

    max_price = numeric_query_param(:max_price)
    return if performed?

    products = products.where("price >= ?", min_price) unless min_price.nil?
    products = products.where("price <= ?", max_price) unless max_price.nil?

    render json: products.map { |product| serialize_product(product) }
  end

  def show
    render json: serialize_product(Product.find(params[:id]))
  end

  private

  def numeric_query_param(name)
    value = params[name]
    return nil if value.nil?
    return value.to_f if value.to_s.match?(/\A\d+(\.\d+)?\z/) && value.to_f.positive?

    render_error("Query parameter '#{name}' is invalid", :unprocessable_entity)
    nil
  end

  def serialize_product(product)
    {
      id: product.id,
      name: product.name,
      price: product.price
    }
  end
end
